### ISSUE TITLE
Concurrently npm package appears to not be compatible

### ISSUE THREAD

**Issue by xRiot** [KnapsackPro/knapsack-pro-cypress#20]
It seems like we can't run Cypress tests simultaneously using the NPM package [Concurrently ](https://www.npmjs.com/package/concurrently) v5.0.0 since as soon as the first test starts in both workers, the package exists with no error and the message is: `[2]+  Stopped`

We're trying to use the following command _(and rest of the environment variables are in our .env file)_:
```
yarn concurrently "KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=0 knapsack-pro-cypress" "KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=1 knapsack-pro-cypress"
```
It works fine when using two shells or Jenkins.

**Comment by ArturT** [comment]
Is it working fine when you run knapsack-pro-cypress in two shells at the same time? 

I'm guessing maybe a problem with using `Concurrently` is actually on Cypress side not being able to run 2 Cypress processes at the same time.

**Comment by xRiot** [comment]
Yes that's correct @ArturT, it works fine if we open two shells and also with our Jenkins pipeline.

Are there any way we can address this conflict or steps we can try?

**Comment by ArturT** [comment]
This is how I run Cypress under the hood:
https://github.com/KnapsackPro/knapsack-pro-cypress/blob/b32f350dde5653a0ecc6c9d681f4f638221e6b8c/src/knapsack-pro-cypress.ts#L37

https://docs.cypress.io/guides/guides/command-line.html#cypress-run

Maybe this will help you debug that.

**Comment by xRiot** [comment]
Hello @ArturT.

For clarification, did you manage to run the parallel tests with npm package concurrency by awaiting `cypress.run` on your end?

Thanks.

**Comment by ArturT** [comment]
Can you help with that? You can fork this example repo and adjust bash script to use `Concurrently`
package https://github.com/KnapsackPro/cypress-example-kitchensink/blob/knapsack-pro/bin/knapsack_pro_cypress

**Comment by xRiot** [comment]
I will try to get back to you next week with some results, thanks for the patience and help!

**Comment by xRiot** [comment]
Thank you for your patience. It seems like all the tests are working in your repository with concurrency since they're running just fine and passing @ArturT.

Although I've observed that this setup might not be perfect just yet _(perhaps more tweaking is required)_, firstly the output of the tests are colliding in the terminal and lastly the test number is always wrong. For example, it says `Running: xx (1 of 1)` while it's not at the first test and naturally there's more than 1 test. On the bright side, in _our_ repository the test number is shown correctly _(even though cypress quits before the tests run)_.

If anyone wants to replicate this:
```
# Clone https://github.com/KnapsackPro/cypress-example-kitchensink
npm install
npm start
npx concurrently "KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=0 KNAPSACK_PRO_CI_NODE_BUILD_ID=VybC473yLXCBeeE2e98nuE6ZGZFcMmQ5 knapsack-pro-cypress" "KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=1 KNAPSACK_PRO_CI_NODE_BUILD_ID=hMrrzUtud8V5PcD5v3FS92x5zeMS89JZ knapsack-pro-cypress"
```
Remember to also set `KNAPSACK_PRO_ENDPOINT`, `KNAPSACK_PRO_TEST_SUITE_TOKEN_CYPRESS` and if you didn't fork the repository set `KNAPSACK_PRO_COMMIT_HASH`

To conclude this, at the end of the day the major differences between our repository and yours is that we're not awaiting `cypress.run`, we have a more up to date Cypress version and we're using yarn _(which should be irrelevant)_. We'll need to debug this further on our side but from my understanding concurrency should be compatible with the right(?) configuration. I'll try to update this post once we find the reason for Cypress exiting before running the tests in our setup, maybe someone else could also help shine more light on this :thumbsup:

**Comment by ArturT** [comment]
In your example how to run it with `concurrently` you must set `KNAPSACK_PRO_CI_NODE_BUILD_ID` for node 0 and node 1 in order to run tests from the same Queue.

```
# Clone https://github.com/KnapsackPro/cypress-example-kitchensink
npm install
npm start
npx concurrently "KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=0 KNAPSACK_PRO_CI_NODE_BUILD_ID=should_be_the_same knapsack-pro-cypress" \
"KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=1 KNAPSACK_PRO_CI_NODE_BUILD_ID=should_be_the_same knapsack-pro-cypress"
```


> For example, it says Running: xx (1 of 1) while it's not at the first test and naturally there's more than 1 test. 

When tests are not yet recorded on the Knapsack Pro API side then `@knapsack-pro/cypress` takes one by one test file from Queue. So Cypress shows that you run 1 test file.


What version of Cypress do you use? Is it the latest 3.7.0? https://www.npmjs.com/package/cypress

I'll try update Cypress to 3.7.0 in my repo https://github.com/KnapsackPro/cypress-example-kitchensink and see if it works as expected.

**Comment by xRiot** [comment]
Hey again @ArturT.

Thanks, I've adjusted the command according to your observations and you're correct, when the API is reachable or in fallback mode, the tests display the correct step number and count for the worker.

Here's the updated and minified steps if anyone needs it for reference:
<details>
  <summary>Commands spoiler</summary>

```
# Fork https://github.com/KnapsackPro/cypress-example-kitchensink
npm install concurrently wait-on
npx concurrently "npm:start:ci" "wait-on tcp:8080 && KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=0 knapsack-pro-cypress" "wait-on tcp:8080 && KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=1 knapsack-pro-cypress"
```
Additionally set environment variables: `KNAPSACK_PRO_ENDPOINT`, `KNAPSACK_PRO_TEST_SUITE_TOKEN_CYPRESS`, `KNAPSACK_PRO_CI_NODE_BUILD_ID` and if you didn't fork the repository set `KNAPSACK_PRO_COMMIT_HASH`
</details>

That's correct, we're using Cypress 3.7.0 and we tried that version with your repository and it still worked with concurrently. We haven't yet fully investigated the cause of Cypress exiting in our repository, I'll update this as soon as we find the reason.

Although it's good to note that this setup makes the output almost unreadable due to the worker's ouput colliding and furthermore the final Cypress results aren't bundled together, maybe there's a better way to let our devs run Cypress tests locally in parallel than with this approach?

**Comment by ArturT** [comment]
BTW You can change `KNAPSACK_PRO_LOG_LEVEL` to `warn` to show fewer logs output.
https://github.com/KnapsackPro/knapsack-pro-core-js#how-to-change-log-level
You would still have mixed output from parallel cypress instances but at least less output from knapsack pro.

BTW I'm not a fan of running tests in parallel in development like that.

```
npx concurrently "npm:start:ci" "wait-on tcp:8080 && etc...
```

because you start a single http server and at the same time you run 2 Cypress processses tests against this http server.
You need to have tests written in a way that they won't interrupt each other if one test changes state of DB for the app running on the test http server. If you would have hooks that clear DB state then one test could clean up DB while other parallel Cypress is still testing something on the http server.

**Comment by xRiot** [comment]
Thanks, that's some very good points @ArturT and we've noticed this being an issue with our way of performing the tests.

We tried `cypress open` but this doesn't cover the entire scope of the project and because of that our developers prefer to instead use `cypress run` locally but this approach is too time consuming and we figured parallel workers would solve this but instead it created a new layer of complexity. Perhaps if we tweak this correctly we could achieve better results or if there's a different way of doing this locally then we'd like to hear about it.

Also like I mentioned earlier, if we find a better way or a way of improving this then I'll make sure to update this.

**Comment by ArturT** [comment]
I can recommend this style of work:

* developers work on feature branches and run locally tests only related to a part of the app they are testing or creating tests for
* you push often changes to CI where your feature branch is tested (the whole test suite is executed)
* in development on developer machine you never run the whole test suite (it's up to CI to take care of running the whole test suite for you)
* CI runs tests in parallel on isolated CI nodes. There is separate DB for each parallel CI node so thanks to that there are no conflicts between tests executing in parallel

If there are very specific reasons why this workflow would not work in your case and you must run all tests locally let me know.
